### PR TITLE
feat(node): add compile build target for node libs

### LINKED ISSUE #2189: node: support  creating libraries and publishable libraries
# Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation (nx.dev) and found no answer
- [x] I checked to make sure that this issue has not already been filed
- [x] I'm reporting the issue to the correct repository (not related to React, Angular or any dependency)

## Expected Behavior

To officially support `@nx g @nrwl/node:lib my-node-utils --publishable`

We can already generate node library via schematic, although it's not documented (inherited from `@nrwl/workspace` I assume):

image
image

## Current Behavior

Currently I can generate node library via `@nx g @nrwl/node:lib my-node-utils`,
which is missing `build` builder. I can add it on my own ofc, but it would be much better to add `--publishable` functionality.

- version of Nx used (_Please run `nx report` at the root of your workspace and copy the results here if you are using Nx 8.6.1 or greater_): `8.9.0`
- 3rd-party libraries and their versions
- and most importantly - a use-case that fails

### PR DESCRIPTION
**Pull request by Cammisuli**
_[Please make sure you have read the submission guidelines before posting an PR](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#submit-pr)_

> _Please make sure that your commit message follows our format._

> Example: `fix(nx): must begin with lowercase`

## Current Behavior (This is the behavior we have today, before the PR is merged)
Node libs cannot be built independently 
## Expected Behavior (This is the new behavior we can expect after the PR is merged)
Node libs can be built using the `@nwrl/node:compile` build target
## Issue
Closes #2189

### CODE REVIEW

File: packages/node/src/builders/compile/compile.impl.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      }
+    }
+  });
+}
+
+function updatePackageJson(
+  options: NodeCompilerBuilderOptions,
+  context: BuilderContext
+) {
+  const mainFile = relative(dirname(options.main), options.main);
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]:
```suggestion
  const mainFile = basename(options.main);
```

File: packages/node/src/builders/compile/compile.impl.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  const typingsFile = mainFile.replace('.ts', '.d.ts');
+  const mainJsFile = mainFile.replace('.ts', '.js');
+  const packageJson = readJsonFile(
+    join(context.workspaceRoot, options.packageJson)
+  );
+  packageJson.main = `./${mainJsFile}`;
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]: Consider when someone changes where tsconfig.json root is.. We.shouldn't assume people won't do weird stuff :)

File: packages/node/src/builders/compile/compile.impl.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+function copyAssetFiles(
+  options: NodeCompilerBuilderOptions,
+  context: BuilderContext
+) {
+  const outDir = options.outputPath;
+  const files: { input: string; output: string }[] = [];
+
+  options.assets.forEach(asset => {
+    if (typeof asset === 'string') {
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]: Can we support globs as a string?

File: packages/node/src/builders/compile/compile.impl.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+function copyAssetFiles(
+  options: NodeCompilerBuilderOptions,
+  context: BuilderContext
+) {
+  const outDir = options.outputPath;
+  const files: { input: string; output: string }[] = [];
+
+  options.assets.forEach(asset => {
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]: We should normalize this option into 1 data structure

File: packages/node/src/builders/compile/compile.impl.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+      globbedFiles.forEach(globbedFile => {
+        files.push({
+          input: join(context.workspaceRoot, asset.input, globbedFile),
+          output: join(context.workspaceRoot, outDir, asset.output, globbedFile)
+        });
+      });
+    }
+  });
+
+  files.forEach(file => {
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]:
```
Promise.all(
  files.map(f => copy(f.input, f.output))
)
```

File: packages/node/src/schematics/library/library.ts [hunk 6]
```diff
@@ -81,9 +83,12 @@ function createFiles(options: NormalizedSchema): Rule {
         offsetFromRoot: offsetFromRoot(options.projectRoot)
       }),
       move(options.projectRoot),
-      options.unitTestRunner === 'none'
-        ? filter(file => !file.endsWith('spec.ts'))
-        : noop()
+      options.unitTestRunner === 'jest'
```
Comment (FrozenPandaz) [inline packages/node/src/schematics/library/library.ts]: The last one will work when new test runners are added so I don't see a reason to change this.

File: packages/node/src/schematics/library/library.ts [hunk 7]
```diff
@@ -102,3 +107,26 @@ function updateTsConfig(options: NormalizedSchema): Rule {
...
+    const architect = json.projects[options.name].architect;
+    if (architect) {
+      architect.build = {
+        builder: '@nrwl/node:compile',
+        options: {
+          outputPath: `dist/libs/${options.projectDirectory}`,
+          tsConfig: `${options.projectRoot}/tsconfig.lib.json`,
+          packageJson: `${options.projectRoot}/package.json`,
+          main: `${options.projectRoot}/src/index.ts`,
+          assets: [`${options.projectRoot}/README.md`]
```
Comment (FrozenPandaz) [inline packages/node/src/schematics/library/library.ts]: When we support globs.. can we put like...
```
**/*.md
**/*.json
```

File: packages/node/src/builders/compile/compile.impl.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+};
+
+export default createBuilder(run);
+
+function run(options: NodeCompilerBuilderOptions, context: BuilderContext) {
+  return compileTypeScriptFiles(options, context).pipe(
+    tap(() => {
+      updatePackageJson(options, context);
+      copyAssetFiles(options, context);
+    })
```
Comment (FrozenPandaz) [inline packages/node/src/builders/compile/compile.impl.ts]: `context.logger.info('It's done! Here it is: `

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Review by Hotell — COMMENTED** [review]
Amazing! Thanks a lot ❤️

**Pull request merged**